### PR TITLE
#16b990586  ch(databaseSetup manipulatingData)  setup database, versioning and make requests to endpoints

### PR DESCRIPTION
**Pull request by banobepascal**
#### What does this PR do?
- Versions our Api with v2
- setup database to store, manipulate, retrieve dat
- sets endpoints for database

#### Description of Task to be completed?
- completed user signin and signup
- completed endpoint GET all articles posted
- completed endpoint for GET specific article with ID
- completed endpoint for POST an article
- completed enpoint for EDIT an article
- completed endpoint for DELETE article
- completed endpoint for GET  specific articles with tag or title
- completed enpoint for flagging an article as inappropiate

#### How should this be manually tested?
A user can navigate to the branch(ch-databaseSetup-16b990586) and clone the repo then run `npm run dbSetup` to create database

#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/1689905867

### CODE REVIEW

File: api/v2/controllers/adminController.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable padded-blocks */
+/* eslint-disable radix */
+import articles from '../models/article';
+
+class AdminAccess {
+  // delete specific flagged article
```
Comment (NkFab) [inline api/v2/controllers/adminController.js]: There's a better way to comment on methods using Jsdocs format. For instance, you can do rewrite this comment as /**
            * @description delete specific flagged article
            * @param {Object} req
            * @param {Object} res
            * @returns {Object} Returns the response
            / ***

*Note:* Do this on every method where you have comments.

File: api/v2/controllers/articleController.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  static flagArticle(req, res) {
+    const article = articles.find((a) => a.id === parseInt(req.params.id));
+    if (!article) {
+      return res.status(404).json({
+        status: 404,
+        error: 'article not found',
+      });
+    }
+
+    const { error } = validation.validateFlag(req.body);
```
Comment (NkFab) [inline api/v2/controllers/articleController.js]: A method should only do what it was created for, in this case, it was created to flag an article and nothing else, so you should move the validation in its separate folder and apply it in the route as a middleware.

File: api/v2/controllers/commentController.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  static async commentArticle(req, res) {
+    const article = articles.find((a) => a.id === parseInt(req.params.id));
+    if (!article) {
+      return res.status(404).json({
+        status: 404,
+        error: 'article not found',
+      });
+    }
+
+    const { error } = validation.validateComment(req.body);
```
Comment (NkFab) [inline api/v2/controllers/commentController.js]: A method should only do what it was created for, in this case, it was created to create a comment on an article and nothing else, so you should move the validation in its separate folder and apply it in the route as a middleware.

File: api/v2/controllers/userController.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+import jwt from 'jsonwebtoken';
+import users from '../models/users';
+import validation from '../middleware/validation';
+
+ENV.config();
+
+class UserController {
+  // create user account
+  static async signUp(req, res) {
+    const { error } = validation.validateUserSignUp(req.body);
```
Comment (NkFab) [inline api/v2/controllers/userController.js]: A method should only do what it was created for, in this case, it was created to create an account and nothing else, so you should move the validation in its separate folder and apply it in the route as a middleware.

File: api/v2/controllers/userController.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  static async loginUser(req, res) {
+    const checkEmail = users.find((user) => user.email === req.body.email);
+    if (!checkEmail) {
+      return res.status(400).json({
+        status: 400,
+        error: 'wrong email or password',
+      });
+    }
+
+    const validPassword = await bcrypt.compare(
```
Comment (NkFab) [inline api/v2/controllers/userController.js]: This should be moved to a separate helper folder for encrypting passcodes.

File: api/v2/controllers/userController.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+      });
+    }
+
+    const signinPayLoad = {
+      email: checkEmail.email,
+      password: checkEmail.password,
+      isAdmin: checkEmail.isAdmin,
+    };
+
+    const gentoken = jwt.sign(signinPayLoad, process.env.JWT_KEY);
```
Comment (NkFab) [inline api/v2/controllers/userController.js]: The token generation should be moved to a helper function directory.

File: api/v2/middleware/admin.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+/* eslint-disable func-names */
+/* eslint-disable consistent-return */
+
+module.exports = function (req, res, next) {
```
Comment (NkFab) [inline api/v2/middleware/admin.js]: Why are you using es5?

File: api/v2/middleware/prod.js [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import helmet from 'helmet';
+import compression from 'compression';
+
+// eslint-disable-next-line func-names
+module.exports = function (app) {
```
Comment (NkFab) [inline api/v2/middleware/prod.js]: Why are you mixing es5 with es6?

File: api/v2/migrations/dbSetup.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import setupTables from './dbQuery';
+
+const config = {
+  user: 'postgres',
+  database: 'teamworkdb',
+  password: 'postgres',
+  port: 5432,
+  max: 10,
+  idleTimeoutMillis: 30000,
+};
```
Comment (NkFab) [inline api/v2/migrations/dbSetup.js]: This config object data shouldn't be accessible to the public move it to the dotenv file.

File: api/v2/migrations/dropTables.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+import { dropTables } from './dbQuery';
+
+const config = {
+  user: 'postgres',
+  database: 'teamworkdb',
+  password: 'postgres',
+  port: 5432,
+  max: 10,
+  idleTimeoutMillis: 30000,
+};
```
Comment (NkFab) [inline api/v2/migrations/dropTables.js]: This config object data shouldn't be accessible to the public move it to the dotenv file.

File: api/v2/models/article.js [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    createdOn: 'September 20, 2019 8:36 PM',
+    title: 'Freelancing',
+    article: 'Do you want to become a top class freelancer',
+    authorId: author.email,
+    status: false,
+    comments: [],
+  },
+];
+
+export default articles;
```
Comment (NkFab) [inline api/v2/models/article.js]: Is this file still needed? If yes, for what?

File: api/v2/models/comment.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    comment: 'good article',
+  },
+  {
+    comment_id: 2,
+    authorId: author.id + 1,
+    comment: 'awesome article',
+  },
+];
+
+export default comments;
```
Comment (NkFab) [inline api/v2/models/comment.js]: Is this file still needed? If yes, for what?

File: api/v2/models/users.js [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    email: 'jessica@test.com',
+    password: '$2b$10$L2tLrhxqO9CqTJ3shPxKlusN23/pZa/nlAwc3mvAQV9CAIrh5hRgK',
+    gender: 'female',
+    jobRole: 'Backend developer',
+    department: 'Engineering',
+    address: 'Kigali',
+  },
+];
+
+export default users;
```
Comment (NkFab) [inline api/v2/models/users.js]: Is this file still needed? If yes, for what?

File: api/v2/routes/admin.js [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import express from 'express';
+import admin from '../controllers/adminController';
+import adminAuth from '../middleware/admin';
+import validation from '../middleware/validation';
+
+const adminRoute = express.Router();
+
+adminRoute.delete(
+  '/api/v1/flagged/articles/:id',
+  [validation.checkToken, adminAuth],
```
Comment (NkFab) [inline api/v2/routes/admin.js]: Why did you add the brackets on these imported middlewares?

File: api/v2/controllers/adminController.js (hunk 1 above)
Comment (banobepascal) [reply api/v2/controllers/adminController.js]: thanks for the recomendation point noted

File: api/v2/controllers/articleController.js (hunk 2 above)
Comment (banobepascal) [reply api/v2/controllers/articleController.js]: working on it

File: api/v2/models/comment.js (hunk 12 above)
Comment (banobepascal) [reply api/v2/models/comment.js]: thanks for the feedback i am working on it

File: .env [hunk 15]
```diff
@@ -1 +1,3 @@
-JWT_KEY=teamwork
\ No newline at end of file
+JWT_KEY=teamwork
+
+DATABASE_URL=postgres://postgres:postgres@localhost:5432/teamworkdb
```
Comment (NkFab) [inline .env:3]: The .env file shouldn't available to the public because there's no need to have it if it will be available. Git-ignore this file.

File: api/v2/middleware/verifyToken.js [hunk 16]
```diff
@@ -0,0 +1,39 @@
+/* eslint-disable consistent-return */
+import ENV from 'dotenv';
+import jwt from 'jsonwebtoken';
+// import client from '../helpers/dbConnection';
```
Comment (NkFab) [inline api/v2/middleware/verifyToken.js:4]: Remove this commented line.

**Review by NkFab — COMMENTED** [review]
Increase your code coverage

**Comment by banobepascal** [comment]
Hey Manzi i am going to resolve this PR and send you a new PR in 10 minutes with the appropiate git branch convention. since it is affects my Tracker stories
